domain/src/main/java/org/oppia/android/domain/state/StateDeck.kt

```diff
@@ -251,4 +285,37 @@ class StateDeck constructor(
   private fun isTopOfDeckTerminal(): Boolean {
     return isTopOfDeckTerminalChecker(pendingTopState)
   }
+
+  /** Navigates to the old state in the deck if it is possible. */
+  fun revisitOldCard() {
+    if (showLearnAgainButton && revisionIndex != -1) {
+      stateIndex = revisionIndex
+      showLearnAgainButton = false
+      revisionIndex = -1
+    }
+  }
+
+  /**
+   * Checks if the given state is present in the [previousStates] list.
+   * If found, sets [revisionIndex] to the corresponding index.
+   */
+  fun isStatePreviouslyVisited(stateName: String): Boolean {
```

Comment:
- "is" implies present tense, but we are looking for something that happened before, so `wasStateVisitedBefore` would have been more appropriate.

- Conventionally, functions with "isXyz" are generally expected to be pure (without side effects). 
 
- "find" indicates that it's looking for something, so, maybe rename to:

```suggestion
  fun findPreviouslyVisitedState(stateName: String): Boolean {
```

Reply:
This function return boolean value but also sets the revision index. 
Your suggested name also makes sense.

Reply:
> This function return boolean value but also sets the revision index. Your suggested name also makes sense.

True, that's what makes it confusing. I think we can perform a small refactor so that it does not have 2 responsibilities. E.g.:

```kotlin
/**
 * Checks if the given state was previously visited.
 * If found, updates [revisionIndex] and returns true.
 */
fun findPreviouslyVisitedState(stateName: String): Boolean {
    return getStateIndex(stateName)?.let {
        revisionIndex = it
        true
    } ?: false
}

/**
 * Finds the index of the given state in the [previousStates] list.
 * Returns the index if found, or null otherwise.
 */
private fun getStateIndex(stateName: String): Int? {
    for (i in previousStates.size - 1 downTo 0) {
        if (previousStates[i].state.name == stateName) {
            return i
        }
    }
    return null
}
```

Overall, I am still unsure of the boolean function naming `findPreviouslyVisitedState`. It could be `wasStateVisitedBefore()`. I would appreciate third party input. Could you PTAL @seanlip?

The usage context in `ExplorationProgressController` is:

```kotlin
 val hasVisitedStateBefore = explorationProgress.stateDeck.findPreviouslyVisitedState(outcome.destStateName)

 answerOutcome =  explorationProgress.stateGraph.computeAnswerOutcomeForResult(topPendingState, outcome, hasVisitedStateBefore)

```

`StateGraph.computeAnswerOutcomeForResult()` expects a boolean variable: 

```kotlin
  fun computeAnswerOutcomeForResult(
    currentState: State,
    outcome: Outcome,
    wasStateVisitedBefore: Boolean
  ): AnswerOutcome 
```

Reply:
Both findPreviouslyVisitedState and wasStateVisitedBefore don't convey that the function has the side-effect of setting revisionIndex. From only reading the (revised) implementation, I think that that side-effect is more important than the return value and a possible name might be `setRevisionIndexToStateName` or similar.

That said, the usage context does suggest a boolean. If it's important to set things, why don't you have ExplorationProgressController get the state index and do the setting explicitly? Stuff like find/get/etc. is generally not expected to have side effects.

Reply:
Thanks @seanlip! If we have the controller set the index, then things would become clearer:
 
```kotlin
/**  Checks if the given state was previously visited. */
fun wasStateVisitedBefore(stateName: String): Boolean {
    return getStateIndex(stateName) != null
}

/** Sets [revisionIndex] to the index of the given state if found. */
fun setRevisionIndexIfVisited(stateName: String) {
    return getStateIndex(stateName)?.let {
        revisionIndex = it
    }
}

/**
 * Finds the index of the given state in the [previousStates] list.
 * Returns the index if found, or null otherwise.
 */
private fun getStateIndex(stateName: String): Int? {
    for (i in previousStates.size - 1 downTo 0) {
        if (previousStates[i].state.name == stateName) {
            return i
        }
    }
    return null
}
```

Usage in `ExplorationProgressController`:

```kotlin

val wasVisited = explorationProgress.stateDeck.wasStateVisitedBefore(outcome.destStateName)
explorationProgress.stateDeck.setRevisionIndexIfVisited(outcome.destStateName)

answerOutcome = explorationProgress.stateGraph.computeAnswerOutcomeForResult(
    topPendingState, outcome, wasVisited
)
```

Reply:
Thanks @adhiamboperes , It is more readable and clear to understand.

Reply:
Done.